flex_model/distributed/distributed_state.py

```diff
@@ -118,65 +120,211 @@ def build(
         data_parallel_size,
     )
     return device_mesh
 
 
-def _initialize_distributed_groups(device_mesh: GPUDeviceMesh) -> None:
+def _initialize_distributed_groups(
+    rank: int,
+    device_mesh: GPUDeviceMesh,
+) -> None:
     """Initializes the torch distributed groups.
 
     Constructs the torch distributed groups defined in the device mesh
     and saves the corresponding group handles for the local device.
     """
-    rank = dist.get_rank()
-
     # Initialize torch distributed tp groups
-    global _TENSOR_PARALLEL_GROUP
     for group_ranks in device_mesh.tp_group_ranks:
         group = dist.new_group(group_ranks)
         if rank in group_ranks:
-            _TENSOR_PARALLEL_GROUP = group
+            tp_group = group
             logger.debug(
                 f"Torch rank {dist.get_rank()} init TP group: {group_ranks}"
             )
 
     # Initialize torch distributed dp groups
-    global _DATA_PARALLEL_GROUP
     for group_ranks in device_mesh.dp_group_ranks:
         group = dist.new_group(group_ranks)
         if rank in group_ranks:
-            _DATA_PARALLEL_GROUP = group
+            dp_group = group
             logger.debug(
                 f"Torch rank {dist.get_rank()} init DP group: {group_ranks}"
             )
 
     # Initialize torch distributed pp groups
-    global _PIPELINE_PARALLEL_GROUP
     for group_ranks in device_mesh.pp_group_ranks:
         group = dist.new_group(group_ranks)
         if rank in group_ranks:
-            _PIPELINE_PARALLEL_GROUP = group
+            pp_group = group
             logger.debug(
                 f"Torch rank {dist.get_rank()} init PP group: {group_ranks}"
             )
 
+    return tp_group, pp_group, dp_group
+
+
+@dataclass
+class LocalParallelState:
+    local_process_group: dist.ProcessGroup
+    local_rank: int
+    local_world_view: List[int]
+    local_tp_group: dist.ProcessGroup
+    local_pp_group: dist.ProcessGroup
+    local_dp_group: dist.ProcessGroup
+    local_mesh: GPUDeviceMesh
+
+    def __repr__(self):
+        _repr = "LocalParallelState(\n"
+        for field in fields(self):
+            _repr += f"\t{field.name}: {getattr(self, field.name)}\n"
+        _repr += ")"
+        return _repr
+
+
+@dataclass
+class GlobalParallelState:
+    global_world_view: List[int]
+    model_to_subset_state_map: Dict[nn.Module, LocalParallelState] = None
+
+
+def _initialize_global_distributed_state(global_world_view):
+    global _GLOBAL_PARALLEL_STATE
+    _GLOBAL_PARALLEL_STATE = GlobalParallelState(
+        global_world_view=global_world_view,
+        model_to_subset_state_map={},
+    )
+
+    return
+
+
+def _dataclass_from_dict(cls, mapping):
+    cls_fields = {f.name for f in fields(cls) if f.init}
+    arg_fields = {
+        name: value for name, value in mapping.items() if name in cls_fields
+    }
+    return cls(**arg_fields)
+
+
+def _local_to_subset_device_mesh(
+    local_world_view_device_mesh, local_to_subset_view_map
+):
```

Comment:
Probably shouldn't require two instances of `GPUDeviceMesh`s for local and subset world views. Might be cleaner to instantiate one per `FlexModel` instance, where it keeps both local and subset groups in its attributes as opposed to just local.